### PR TITLE
2830 IEC WPP full validation

### PR DESCRIPTION
**Pull request by GerardThevenon**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo [hunk 1]
```diff
@@ -18,7 +18,7 @@ model WPP4ACurrentSource "Wind Power Plant Type 4A model from IEC 61400-27-1 sta
 
   extends Icons.Example;
 
-  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.379, tBegin = 12, tEnd = 12.15) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.4, tBegin = 12, tEnd = 12.15) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:21]: It starts to get a bit high there, what doesn't work with a frank fault ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4BCurrentSource.mo [hunk 2]
```diff
@@ -18,7 +18,7 @@ model WPP4BCurrentSource "Wind Power Plant Type 4A model from IEC 61400-27-1 sta
 
   extends Icons.Example;
 
-  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.379, tBegin = 12, tEnd = 12.15) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.4, tBegin = 12, tEnd = 12.15) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4BCurrentSource.mo:21]: same

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4ACurrentSource.mo [hunk 3]
```diff
@@ -20,7 +20,7 @@ model WT4ACurrentSource "Wind Turbine Type 4A model from IEC 61400-27-1 standard
...
     Placement(visible = true, transformation(origin = {-130, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
-  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.379, tBegin = 12, tEnd = 12.15) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.4, tBegin = 12, tEnd = 12.15) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4ACurrentSource.mo:23]: same

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4BCurrentSource.mo [hunk 4]
```diff
@@ -20,7 +20,7 @@ model WT4BCurrentSource "Wind Turbine Type 4B model from IEC 61400-27-1 standard
...
     Placement(visible = true, transformation(origin = {-130, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
-  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.379, tBegin = 12, tEnd = 12.15) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault1(RPu = 0, XPu = 0.4, tBegin = 12, tEnd = 12.15) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4BCurrentSource.mo:23]: same

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4BCurrentSource.mo [hunk 5]
```diff
@@ -38,7 +38,7 @@ model WT4BCurrentSource "Wind Turbine Type 4B model from IEC 61400-27-1 standard
     Placement(visible = true, transformation(origin = {110, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
   Dynawo.Electrical.Buses.InfiniteBusWithVariations infiniteBusWithVariations(U0Pu = 1.0678, UEvtPu = 0, UPhase = -0.04, omega0Pu = 1, omegaEvtPu = 0.99, tOmegaEvtEnd = 21, tOmegaEvtStart = 20, tUEvtEnd = 0, tUEvtStart = 0) annotation(
     Placement(visible = true, transformation(origin = {140, 0}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
-  Dynawo.Electrical.Events.NodeFault nodeFault(RPu = 0, XPu = 0.091, tBegin = 6, tEnd = 6.4) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault(RPu = 0, XPu = 0.145, tBegin = 6, tEnd = 6.4) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WT4BCurrentSource.mo:41]: same

File: examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par [hunk 6]
```diff
@@ -193,13 +193,13 @@
   </set>
   <set id="Fault">
     <par type="DOUBLE" name="fault_RPu" value="0"/>
-    <par type="DOUBLE" name="fault_XPu" value="0.315"/>
+    <par type="DOUBLE" name="fault_XPu" value="0.38"/>
```
Comment (cossartque) [inline examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par:196]: same

File: examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par [hunk 7]
```diff
@@ -69,7 +69,7 @@
     <par type="DOUBLE" name="WPP_Kpufrt" value="2"/>
     <par type="DOUBLE" name="WPP_Kqv" value="2"/>
     <par type="INT" name="WPP_Mqfrt" value="1"/>
-    <par type="INT" name="WPP_MqG" value="1"/>
+    <par type="INT" name="WPP_MqG" value="2"/>
```
Comment (cossartque) [inline examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par:72]: I suggest that you add several examples for the different configurations of the table

File: examples/DynaSwing/IEC/Wind/Neplan/WT4ACurrentSource/WT4ACurrentSource.par [hunk 8]
```diff
@@ -157,13 +157,13 @@
   </set>
   <set id="Fault">
     <par type="DOUBLE" name="fault_RPu" value="0"/>
-    <par type="DOUBLE" name="fault_XPu" value="0.315"/>
+    <par type="DOUBLE" name="fault_XPu" value="0.36"/>
```
Comment (cossartque) [inline examples/DynaSwing/IEC/Wind/Neplan/WT4ACurrentSource/WT4ACurrentSource.par:160]: same

File: examples/DynaSwing/IEC/Wind/Neplan/WT4BCurrentSource/WT4BCurrentSource.par [hunk 9]
```diff
@@ -162,13 +162,13 @@
   </set>
   <set id="Fault">
     <par type="DOUBLE" name="fault_RPu" value="0"/>
-    <par type="DOUBLE" name="fault_XPu" value="0.315"/>
+    <par type="DOUBLE" name="fault_XPu" value="0.36"/>
```
Comment (cossartque) [inline examples/DynaSwing/IEC/Wind/Neplan/WT4BCurrentSource/WT4BCurrentSource.par:165]: same

**Review by cossartque — CHANGES_REQUESTED** [review]

**Comment by cossartque** [comment]
And don't forget to check the nrt and update the references when necessary

File: examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par (hunk 7 above)
Comment (GerardThevenon) [reply examples/DynaSwing/IEC/Wind/Neplan/WPP4ACurrentSource/WPP4ACurrentSource.par:72]: How would you do that?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo (hunk 1 above)
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:21]: At the end of the fault, argument of sqrt is < 0

Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:21]: Ok but does the model itself prevent IMax² - ipcmd² < 0 ? Is it just a solver issue or a model issue ?

Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:21]: Yes, with max(0, IMax² - ipcmd²), so solver issue

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/AbsLimRateLimFeedthroughFreezeLimDetection.mo [hunk 10]
```diff
@@ -27,9 +27,9 @@ block AbsLimRateLimFeedthroughFreezeLimDetection "First order feed-through with
     Placement(visible = true, transformation(origin = {-50, 40}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
   Modelica.Blocks.Logical.LessThreshold lessThreshold(threshold = YMin) annotation(
     Placement(visible = true, transformation(origin = {-50, -40}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
-  Dynawo.NonElectrical.Blocks.NonLinear.FixedBooleanDelay fixedBooleanDelay(Y0 = false, tDelay = tS) annotation(
+  Dynawo.NonElectrical.Blocks.NonLinear.FixedBooleanDelay fixedBooleanDelay(Y0 = Y0 > YMax, tDelay = tS) annotation(
```
Comment (GerardThevenon) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/AbsLimRateLimFeedthroughFreezeLimDetection.mo:30]: I anticipate that having a U0 instead of a Y0 could be better (Y0 = U0 > YMax), what do you think @cossartque ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/BaseClasses/BaseWPP.mo [hunk 11]
```diff
@@ -237,11 +237,11 @@ model BaseWPP "Base model for Wind Power Plants from IEC 61400-27-1 standard"
   //Input variables
   Modelica.Blocks.Interfaces.RealInput omegaRefPu(start = SystemBase.omegaRef0Pu) "Reference frame for grid angular frequency in pu (base omegaNom)" annotation(
     Placement(visible = true, transformation(origin = {-140, -40}, extent = {{-20, -20}, {20, 20}}, rotation = 0), iconTransformation(origin = {-110, -60}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
-  Modelica.Blocks.Interfaces.RealInput PPDRefPu(start = -P0Pu * SystemBase.SnRef / SNom) "Active power reference at grid terminal in pu (base SNom) (generator convention)" annotation(
+  Modelica.Blocks.Interfaces.RealInput PWPRefPu(start = -P0Pu * SystemBase.SnRef / SNom) "Reference active power in pu (base SNom) (generator convention)" annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/BaseClasses/BaseWPP.mo:240]: Why this change of name ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4BCurrentSource.mo [hunk 12]
```diff
@@ -18,7 +18,7 @@ model WPP4BCurrentSource "Wind Power Plant Type 4A model from IEC 61400-27-1 sta
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4BCurrentSource.mo:18]: This test case doesn't work

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/AbsLimRateLimFeedthroughFreezeLimDetection.mo (hunk 10 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/AbsLimRateLimFeedthroughFreezeLimDetection.mo:30]: Yes of course, U0, not Y0

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/BaseControls/WT/CurrentLimiter.mo [hunk 13]
```diff
@@ -69,7 +69,7 @@ model CurrentLimiter "Current limitation module for wind turbines (IEC N°61400-
     Placement(visible = true, transformation(origin = {140, 70}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
   Dynawo.NonElectrical.Blocks.NonLinear.MultiSwitch switch1(nu = 3) annotation(
     Placement(visible = true, transformation(origin = {110, 100}, extent = {{-10, 10}, {10, -10}}, rotation = 0)));
-  Modelica.Blocks.Math.Max max3 annotation(
```
[the remark this answers is not in the document]
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/BaseControls/WT/CurrentLimiter.mo:72]: It has always been "min" but a few weeks ago, I noticed that test C was working thanks to "max". However, now that I have checked more curves, changing to "max" was clearly a mistake.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/BaseClasses/BaseWPP.mo (hunk 11 above)
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/BaseClasses/BaseWPP.mo:240]: My mistake, wasn't consistent with the norm.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/WPP/WPP4CurrentSource_INIT.mo [hunk 14]
```diff
@@ -17,10 +17,13 @@ model WPP4CurrentSource_INIT "Wind Power Plant Type 4 model from IEC 61400-27-1
...
   extends Dynawo.Electrical.Wind.IEC.WT.WT4CurrentSource_INIT;
 
-  Dynawo.Types.PerUnit X0Pu "Initial reactive power or voltage reference at grid terminal in pu (base SNom or UNom) (generator convention)";
+  //QControl parameter
+  parameter Integer MwpqMode "Control mode (0 : reactive power reference, 1 : power factor reference, 2 : UQ static, 3 : voltage control)";
+
+  Dynawo.Types.PerUnit X0Pu "Initial reactive power or voltage reference in pu (base SNom or UNom) (generator convention)";
 
 equation
-  X0Pu = XWT0Pu;
```
[the remark this answers is not in the document]
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/WPP/WPP4CurrentSource_INIT.mo:23]: Yes, but it is true that if WT_MqG always is 2 within WPP, XWT0Pu could simply be -Q0Pu * SystemBase.SnRef / SNom

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/BaseConverters/GenSystem4.mo [hunk 15]
```diff
@@ -63,6 +63,7 @@ model GenSystem4 "Type 4 generator system module (IEC N°61400-27-1)"
     Placement(visible = true, transformation(origin = {-150, 20}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-110, 0}, extent = {{10, -10}, {-10, 10}}, rotation = 180)));
   Modelica.Blocks.Interfaces.RealInput iqMinPu(start = IqMin0Pu) "Minimum reactive current at converter terminal in pu (base UNom, SNom) (generator convention)" annotation(
     Placement(visible = true, transformation(origin = {-150, -20}, extent = {{-10, 10}, {10, -10}}, rotation = 0), iconTransformation(origin = {-110, -80}, extent = {{10, -10}, {-10, 10}}, rotation = 180)));
+  input Boolean running(start = true) "Boolean assessing if the converter is connected to the grid";
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/BaseConverters/GenSystem4.mo:66]: "True if the converter is running" (it can be connected and not running :p )

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo [hunk 16]
```diff
@@ -73,6 +75,9 @@ model WT4Injector "Converter model and grid interface according to IEC N°61400-
   Modelica.ComplexBlocks.Interfaces.ComplexOutput uWtPu(re(start = u0Pu.re), im(start = u0Pu.im)) "Complex voltage at grid terminal in pu (base UNom)" annotation(
     Placement(visible = true, transformation(origin = {110, -80}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {110, 60}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
 
+  Types.ActivePowerPu PGenPu(start = -P0Pu) "Active power at WTT in pu (base SnRef) (generator convention)";
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo:78]: WTT ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo [hunk 17]
```diff
@@ -73,6 +75,9 @@ model WT4Injector "Converter model and grid interface according to IEC N°61400-
...
+  Types.ReactivePowerPu QGenPu(start = -Q0Pu) "Reactive power at WTT in pu (base SnRef) (generator convention)";
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo:79]: WTT ? at terminal ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/WT/WT4ACurrentSource.mo [hunk 18]
```diff
@@ -58,8 +61,8 @@ equation
     Line(points = {{-110, -60}, {-92, -60}, {-92, -52}, {-82, -52}}, color = {0, 0, 127}));
   connect(tanPhi, control4A.tanPhi) annotation(
     Line(points = {{-110, -40}, {-82, -40}}, color = {0, 0, 127}));
-  connect(omegaRefPu, control4A.omegaGenPu) annotation(
-    Line(points = {{110, 60}, {60, 60}, {60, 48}, {-60, 48}, {-60, -18}}, color = {0, 0, 127}));
+  connect(const.y, control4A.omegaGenPu) annotation(
```
[the remark this answers is not in the document]
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Wind/IEC/WT/WT4ACurrentSource.mo:62]: Well then can't it just be an input with a start value and optional extvar ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo [hunk 19]
```diff
@@ -36,20 +36,23 @@ model WPP4ACurrentSource "Wind Power Plant Type 4A model from IEC 61400-27-1 sta
...
     Placement(visible = true, transformation(origin = {140, 0}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
-  Dynawo.Electrical.Events.NodeFault nodeFault(RPu = 0, XPu = 0.091, tBegin = 6, tEnd = 6.4) annotation(
+  Dynawo.Electrical.Events.NodeFault nodeFault(RPu = 0, XPu = 0.145, tBegin = 6, tEnd = 6.4) annotation(
     Placement(visible = true, transformation(origin = {50, -40}, extent = {{10, -10}, {-10, 10}}, rotation = 180)));
   Modelica.Blocks.Sources.Step tanPhi(height = 0, offset = -0.21, startTime = 0) annotation(
     Placement(visible = true, transformation(origin = {-170, 60}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
   Modelica.Blocks.Sources.Step xRefPu(height = 0.41, offset = -0.21, startTime = 4) annotation(
     Placement(visible = true, transformation(origin = {-170, -20}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
-  Modelica.Blocks.Sources.Step omegaRefPu(height = 0, offset = 1, startTime = 0) annotation(
+  Modelica.Blocks.Sources.Pulse omegaRefPu(amplitude = -0.01, nperiod = 1, offset = 1, period = 2, startTime = 20) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:45]: Why a pulse ?

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/VariableLimiter.mo [hunk 20]
```diff
@@ -25,7 +25,7 @@ model VariableLimiter "Limit the range of a signal with variable limits"
 
 equation
 
-  y = if u > limit1 then limit1 else if u < limit2 then limit2 else u;
+  y = smooth(0, if u > limit1 then limit1 else if u < limit2 then limit2 else u);
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/VariableLimiter.mo:28]: If it is used elsewhere then we should check all the nrt

**Review by cossartque — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo (hunk 19 above)
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Wind/IEC/Neplan/WPP4ACurrentSource.mo:45]: I did not manage to use Dynawo.Electrical.Controls.Basics.DoubleStep within OpenModelica. The pulse is a perfect fit in my opinion.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo (hunk 17 above)
Comment (GerardThevenon) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/IEC/WT4Injector.mo:79]: Yes, WTT means Wind Turbine Terminal. Is "terminal" enough?

**Review by cossartque — APPROVED** [review]
